rfc/rfc-0000-allow-resources-to-report-status.md

```diff
@@ -0,0 +1,335 @@
+- What use cases does it support?
+  - This feature will enable developers to learn about the progression through the blueprint by looking at the owner status alone. It will also surface information about the health of the underlying resources directly on the owner.
+  - This will also enable integrations that may be looking to create a UI around Cartographer, allowing the UI to display the health of the blueprint without having to make an API call to k8s to retrieve the status of every resource, and without having to know how to define health for every stamped object type.
+- What is the expected outcome?
+  - Users will have more insight into the progression of the blueprint and the health of each underlying resource.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes creating a new CRD, `ClusterResourceStatusRule`. Template authors use this CRD to define the conditions or fields on a stamped resource that determine whether the resource is succeeding or failing. The template author then references this `ClusterResourceStatusRule` from their template.
+
+Cartographer will attempt to read the specified conditions and/or fields and reflect this state on the owner status. If Cartographer cannot determine success or failure, it will show the state as unknown.
+
+```yaml
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterResourceStatusRule
+metadata:
+  name: ready-rule
+spec:
+  healthy:
+    matchConditions:
+      - type: Ready
+        status: 'True'
+  unhealthy:
+    matchConditions:
+      - type: Ready
+        status: 'False'
+
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate
+metadata:
+  name: source
+spec:
+  urlPath: .status.artifact.url
+  revisionPath: .status.artifact.revision
+
+  resourceStatusRef:
+    name: ready-rule
+
+  template:
+    apiVersion: source.toolkit.fluxcd.io/v1beta1
+    kind: GitRepository
+    metadata:
+      name: $(workload.metadata.name)$
+    spec:
+      ...
+    
+---
+apiVersion: carto.run/v1alpha1
+kind: Workload
+metadata:
+  ...
+spec:
+  ...
+status:
+  conditions:
+    ...
+  resources:
+    - name: source-provider
+      outputs:
+        ...
+      stampedRef:
+        ...
+      templateRef:
+        apiVersion: carto.run/v1alpha1
+        kind: ClusterSourceTemplate
+        name: source
+      stampedStatus:
+        status: 'Healthy'
+        conditions:
+          - type: Ready
+            status: 'True'
+            message: 'Fetched revision: main/...'
+            reason: GitOperationSucceeded
+            lastTransitionTime: "2022-03-23T13:08:22Z"
+    - name: source-provider-2
+      outputs:
+        ...
+      stampedRef:
+        ...
+      templateRef:
+        apiVersion: carto.run/v1alpha1
+        kind: ClusterSourceTemplate
+        name: source
+      stampedStatus:
```

Comment:
```suggestion
      stamped:
```

Redundant terminology. this maps:

```
  stampedStatus.status -> stamped.status
  stampedStatus.conditions -> stamped.conditions
```

Reply:
I don't really like `stampedStatus`  as the top level field either, but we already have `stampedRef` as a top level field so I wouldn't want to make another field `stamped`.

Reply:
@squeedee is this still relevant. We are now proposing resources.conditions.

Reply:
Not in the slightest